src/compas_fab/datastructures/partial_order.py

```diff
@@ -0,0 +1,349 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import threading
+from collections import OrderedDict
+from copy import deepcopy
+from itertools import count
+
+import compas
+from compas.data import Data
+from compas.datastructures import Datastructure
+from compas.datastructures import Graph
+
+__all__ = [
+    'Action',
+    'DependencyIdException',
+    'IntegerIdGenerator',
+    'PartialOrder',
+]
+
+
+class IntegerIdGenerator(Data):
+    """Generator object yielding integers sequentially in a thread safe manner.
+
+    Parameters
+    ----------
+    start_value : :obj:`int`
+        First value to be yielded by the generator.
+    """
+    def __init__(self, start_value=1):
+        super(IntegerIdGenerator, self).__init__()
+        self.last_generated = start_value - 1
+        self._lock = threading.Lock()
+        self._generator = count(start_value)
+
+    def __next__(self):
+        with self._lock:
+            self.last_generated = next(self._generator)
+            return self.last_generated
+
+    # alias for ironpython
+    next = __next__
+
+    @property
+    def data(self):
+        return {
+            'start_value': self.last_generated + 1
+        }
+
+    def to_data(self):
+        return self.data
+
+    @classmethod
+    def from_data(cls, data):
+        return cls(data['start_value'])
+
+    @classmethod
+    def from_json(cls, filepath):
+        data = compas.json_load(filepath)
+        return cls.from_data(data)
+
+    def to_json(self, filepath, pretty=False):
+        compas.json_dump(self.data, filepath, pretty=pretty)
+
+
+class DependencyIdException(Exception):
+    """Indicates invalid ids given as dependencies."""
+    def __init__(self, invalid_ids, pa_id=None):
+        message = self.compose_message(invalid_ids, pa_id)
+        super(DependencyIdException, self).__init__(message)
+
+    @staticmethod
+    def compose_message(invalid_ids, pa_id):
+        if pa_id:
+            return 'Planned action {} has invalid dependency ids {}'.format(pa_id, invalid_ids)
+        return 'Found invalid dependency ids {}'.format(invalid_ids)
+
+
+class PartialOrder(Datastructure):
+    """Data structure extending :class:`compas.datastructures.Graph` for creating
+    and maintaining a partially ordered plan (directed acyclic graph).
+    The content of any event of the plan is contained in an
+    :class:`compas_fab.datastructures.Action`.  The dependency ids of a planned
+    action can be thought of as pointers to the parents of that planned action.
+    While actions can be added and removed using the methods of
+    :attr:`compas_fab.datastructures.PartialOrder.graph`, it is strongly recommended
+    that the methods ``plan_action``, ``append_action`` and ``remove_action``
```

Comment:
do you mean ``get_action`` instead of ``plan_action``?

Another thought is, how about renaming ``PartialOrder.get_action()`` to ``PartialOrder.action()``? Is there some meaning associated with the `get_` prefix?

Reply:
Thanks for clarifying, I just somehow had this idea where if there is a `get_` prefix, then I should not modify the returned result in place but rather use the corresponding `set_` method. But I could be hallucinating with other language flavours.